smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/Endpoint.java

```diff
@@ -1,91 +1,75 @@
  */
 
 package software.amazon.smithy.rulesengine.language;
 
-import static software.amazon.smithy.rulesengine.language.error.RuleError.context;
-
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.Comparator;
-import java.util.LinkedHashMap;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.TreeMap;
-import java.util.stream.Collectors;
 import software.amazon.smithy.model.FromSourceLocation;
 import software.amazon.smithy.model.SourceLocation;
 import software.amazon.smithy.model.node.ArrayNode;
 import software.amazon.smithy.model.node.Node;
 import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.StringNode;
 import software.amazon.smithy.model.node.ToNode;
 import software.amazon.smithy.rulesengine.language.error.RuleError;
-import software.amazon.smithy.rulesengine.language.eval.Scope;
-import software.amazon.smithy.rulesengine.language.eval.Type;
-import software.amazon.smithy.rulesengine.language.eval.TypeCheck;
+import software.amazon.smithy.rulesengine.language.evaluation.Scope;
+import software.amazon.smithy.rulesengine.language.evaluation.TypeCheck;
+import software.amazon.smithy.rulesengine.language.evaluation.type.Type;
 import software.amazon.smithy.rulesengine.language.syntax.Identifier;
-import software.amazon.smithy.rulesengine.language.syntax.expr.Expression;
-import software.amazon.smithy.rulesengine.language.syntax.expr.Literal;
-import software.amazon.smithy.rulesengine.language.util.MandatorySourceLocation;
-import software.amazon.smithy.rulesengine.language.util.SourceLocationTrackingBuilder;
-import software.amazon.smithy.rulesengine.language.util.StringUtils;
+import software.amazon.smithy.rulesengine.language.syntax.expressions.Expression;
+import software.amazon.smithy.rulesengine.language.syntax.expressions.literal.Literal;
 import software.amazon.smithy.utils.BuilderRef;
 import software.amazon.smithy.utils.MapUtils;
-import software.amazon.smithy.utils.Pair;
 import software.amazon.smithy.utils.SmithyBuilder;
 import software.amazon.smithy.utils.SmithyUnstableApi;
+import software.amazon.smithy.utils.StringUtils;
 import software.amazon.smithy.utils.ToSmithyBuilder;
 
 /**
- * An Endpoint as returned by EndpointRules.
+ * An EndpointType as returned by EndpointRules.
  */
 @SmithyUnstableApi
-public final class Endpoint extends MandatorySourceLocation implements ToSmithyBuilder<Endpoint>, TypeCheck, ToNode {
+public final class Endpoint implements FromSourceLocation, ToNode, ToSmithyBuilder<Endpoint>, TypeCheck {
     private static final String URL = "url";
     private static final String PROPERTIES = "properties";
     private static final String HEADERS = "headers";
     private static final String SIGV_4 = "sigv4";
     private static final String SIG_V4A = "sigv4a";
     private static final String SIGNING_REGION = "signingRegion";
```

Comment:
Why are these here? Could this just be a TypedPropertiesBag or have some other method of extending what properties an endpoint can have